File: Project4/Ex4.py
import collections

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        # Danh sách kề là dạng biểu diễn chính
        self.adj = [[] for _ in range(vertices)]

    def add_edge(self, u, v):
        """Thêm cạnh vô hướng"""
        self.adj[u].append(v)
        self.adj[v].append(u)

    def to_adjacency_matrix(self):
        """Chuyển đổi sang Ma trận kề"""
        matrix = [[0] * self.V for _ in range(self.V)]
        for u in range(self.V):
            for v in self.adj[u]:
                matrix[u][v] += 1
        return matrix

    def to_adjacency_map(self):
        """Chuyển đổi sang Ánh xạ kề (dictionary)"""
        adj_map = {i: self.adj[i] for i in range(self.V) if self.adj[i]}
        return adj_map

    def print_adj_list(self):
        print("Danh sach ke:")
        for i in range(self.V):
            print(f"{i}: {self.adj[i]}")

class Tree(Graph):
    def to_parent_array(self, root):
        """Chuyển đổi sang Mảng Cha, yêu cầu một đỉnh gốc"""
        parent = [-1] * self.V  # -1 cho gốc và các đỉnh không kết nối
        visited = {root}
        queue = collections.deque([root])

        while queue:
            u = queue.popleft()
            for v in self.adj[u]:
                if v not in visited:
                    visited.add(v)
                    parent[v] = u
                    queue.append(v)
        return parent

if __name__ == "__main__":
    print("--- Bieu dien Do thi ---")
    g = Graph(4)
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 2)
    g.add_edge(2, 3)
    g.add_edge(1, 2)  # Thêm cạnh song song

    g.print_adj_list()
    
    matrix = g.to_adjacency_matrix()
    print("\nMa tran ke:")
    for row in matrix:
        print(row)

    print("\n--- Bieu dien Cay ---")
    t = Tree(5)
    t.add_edge(0, 1)
    t.add_edge(0, 2)
    t.add_edge(1, 3)
    t.add_edge(1, 4)

    t.print_adj_list()

    parent_array = t.to_parent_array(0)
    print("\nMang Cha (goc la 0):")
    for i, p in enumerate(parent_array):
        print(f"parent[{i}] = {p}")

File: Project4/Ex4.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>

using namespace std;

// Lớp biểu diễn đồ thị
class Graph {
public:
    int V; // Số đỉnh
    vector<vector<int>> adj; // Danh sách kề

    Graph(int vertices) : V(vertices), adj(vertices) {}

    // Thêm cạnh (cho đồ thị vô hướng)
    void addEdge(int u, int v) {
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    // Chuyển đổi sang Ma trận kề
    vector<vector<int>> toAdjacencyMatrix() {
        vector<vector<int>> matrix(V, vector<int>(V, 0));
        for (int u = 0; u < V; ++u) {
            for (int v : adj[u]) {
                matrix[u][v]++;
            }
        }
        return matrix;
    }

    // Chuyển đổi sang Ánh xạ kề
    map<int, vector<int>> toAdjacencyMap() {
        map<int, vector<int>> adjMap;
        for (int i = 0; i < V; ++i) {
            if (!adj[i].empty()) {
                adjMap[i] = adj[i];
            }
        }
        return adjMap;
    }

    void printAdjList() {
        cout << "Danh sach ke:" << endl;
        for (int i = 0; i < V; ++i) {
            cout << i << ": ";
            for (int neighbor : adj[i]) {
                cout << neighbor << " ";
            }
            cout << endl;
        }
    }
};

// Lớp biểu diễn cây (kế thừa từ Graph)
class Tree : public Graph {
public:
    Tree(int vertices) : Graph(vertices) {}

    // Chuyển đổi sang Mảng Cha
    vector<int> toParentArray(int root) {
        vector<int> parent(V, -1); // -1 cho gốc và các đỉnh không kết nối
        vector<bool> visited(V, false);
        queue<int> q;

        q.push(root);
        visited[root] = true;

        while (!q.empty()) {
            int u = q.front();
            q.pop();

            for (int v : adj[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    parent[v] = u;
                    q.push(v);
                }
            }
        }
        return parent;
    }
};

int main() {
    cout << "--- Bieu dien Do thi ---" << endl;
    Graph g(4);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 3);
    g.addEdge(1, 2); // Thêm cạnh song song

    g.printAdjList();

    vector<vector<int>> matrix = g.toAdjacencyMatrix();
    cout << "\nMa tran ke:" << endl;
    for (int i = 0; i < g.V; ++i) {
        for (int j = 0; j < g.V; ++j) {
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }
    
    cout << "\n--- Bieu dien Cay ---" << endl;
    Tree t(5);
    t.addEdge(0, 1);
    t.addEdge(0, 2);
    t.addEdge(1, 3);
    t.addEdge(1, 4);

    t.printAdjList();
    
    vector<int> parent_array = t.toParentArray(0);
    cout << "\nMang Cha (goc la 0):" << endl;
    for (int i = 0; i < t.V; ++i) {
        cout << "parent[" << i << "] = " << parent_array[i] << endl;
    }

    return 0;
}